WordPress/gutenberg
# CSS transformation: Fix mangled selectors by switching to a different postcss plugin for prefixing

Fixes  #63829

## What?
Tries switching from `postcss-prefixwrap` to `postcss-prefix-selector` for transforming styles

## Why?
`postcss-prefixwrap` is currently mangling compound `:where` selectors - [Non-iframed Editor: Compound selectors within :where are not prefixed correctly](https://github.com/WordPress/gutenberg/issues/63829#top)

I've reported it upstream in https://github.com/dbtedman/postcss-prefixwrap/issues/515 and can try making a PR on the repo if one hasn't by the time I get to it.

The issue still needs to be fixed in Gutenberg and core, so this PR should do that.

## How?
Switches from [postcss-prefixwrap](https://github.com/dbtedman/postcss-prefixwrap) to [postcss-prefix-selector](https://github.com/RadValentin/postcss-prefix-selector). Why does this make a difference?

I had a look at the codebases for the two. `postcss-prefixwrap` seems to take the whole selector string, and then call the string `split( ',' )` method to handle compound selectors:
https://github.com/dbtedman/postcss-prefixwrap/blob/c2636cdaa8381741df1b2b553d15a82d1e674802/src/internal/domain/CSSRuleWrapper.ts#L20-L46

This results in the selector being split incorrectly and prefixing within a `where` selector.

`postcss-prefix-selector` on the other hand seems to use the `selectors` property provided by the AST which I'd expect to be a lot more advanced in how it splits a selector (it should align pretty closely with the CSS spec):
https://github.com/RadValentin/postcss-prefix-selector/blob/86cb9b9dd50caef6e075ac88f9d716975c68a597/index.js#L38

There are a couple of features not offered by the new dependency:
- Automattic replacing of 'root' selectors rather than prefixing (html, :root, body)
- Avoiding duplicated selectors

It does have a `transform` callback that I've used to implement these features so that there's parity with `trunk`.

## Testing Instructions
1. Activate TT4 (which has button styles) and navigate to the post editor
2. Ensure the editor is not iframed. This can be done by enabling the Custom Fields preference via the top right menu More > Preferences > Custom Fields, then reloading the editor.
3. Edit a post and add some buttons, then save
4. Inspect the buttons and find the matched style with the following selector:
`.editor-styles-wrapper :where(.wp-element-button, .editor-styles-wrapper .wp-block-button__link)`
5. Note the injection of the second `.editor-styles-wrapper` within that


## Screenshots or screencast
